### ISSUE TITLE
Mutator runs twice

### ISSUE THREAD

**Issue by danielbrodin** [jacksleight/statamic-bard-mutator#8]
Hi,

I'm using the mutator to add an id to some headings, which works fine just like in the provided example. However, a heading can realistically exist twice in a document, so would then need to have `id="heading"` and `id="heading-2`. I have kinda solved this, but for some reason the mutator runs twice for each heading so I get `id="heading` and `id="heading-3"`.

I have solved it like this
```php
Mutator::tag('heading', function ($tag, $data, $meta) {
            if ($data->attrs->level === 2 || $data->attrs->level === 3) {
                $slug = str_slug(collect($data->content)->implode('text', ''));
                $this->slugs->push($slug);
                $occurences = $this->slugs->filter(fn ($value) => $value === $slug)->count();
                
                if ($occurences > 1) {
                    $slug = $slug . "-" . $occurences;
                }
                info($slug . " - " . $occurences);
                
                $tag[0]['attrs']['id'] = $slug;
            }

            return $tag;
        });
```

If I give it this content
```html
<h2>Heading</h2>
<h2>Other</h2>
<h2>Heading</h2>
```

The results will be
```html
<h2 id="heading">Heading</h2>
<h2 id="other">Other</h2>
<h2 id="heading-3">Heading</h2>
```

And the log will show
```
heading - 1  
heading-2 - 2  
other - 1  
other-2 - 2  
heading-3 - 3  
heading-4 - 4  
```

Is there any reason for running it twice, or have I maybe missed something?

**Comment by jacksleight** [comment]
Yeah it runs twice because of the way the ProseMirror renderer works. That generates the tag data twice: when the opening tag is rendered and again for the closing tag. Mutator needs to follow what ProseMirror does because if you modified the tag name that would also affect the closing tag.

However, the `$data` object will be the same for both calls, so you could use that to track whether it's the first or second call, and only increment your counter on the first. Something like:

```php
$count = 0;
Mutator::tag('heading', function ($tag, $data, $meta) use ($count) {
    if ($data->attrs->level === 2 || $data->attrs->level === 3) {
        $data->count = $data->count ?? $count++;
        $tag[0]['attrs']['id'] = "heading-{$data->count}";
    }
    return $tag;
});
```

Maybe I should add a flag to the metadata to indicate whether it's the opening or closing tag, could be useful?

**Comment by danielbrodin** [comment]
Ok, that makes sense then. If it is safe to update `$data` then it would work for me to do something like `$data->hasAddedId = true` and do an early return if `$data->hasAddedId` is set.

Knowing if it's the starting or closing tag would be nicer though, and probably usable in other cases.

**Comment by jacksleight** [comment]
Yeah you can update those objects, just dont overwrite any of the standard ProseMirror keys. They only exist inside the the rendering process so altering them wont affect your original data.

The data mutator feature is actually intended for altering the data objects, which you could use instead to generate your ID values (they only run once), then just use the tag mutator to apply it to the tag.

> Knowing if it's the starting or closing tag would be nicer though, and probably usable in other cases.

Now you've got me thinking about it it should actually be possible to have them run once but still provide the data to ProseMirror twice, I'll look into that.

**Comment by jacksleight** [comment]
Fix for this release in 1.1.2.

**Comment by danielbrodin** [comment]
Great 👍 , even nicer than getting the option to know if it's the start or ending tag.
